posts/frangeris/03.secure-values-by-private-keys-in-aws.md

```diff
@@ -0,0 +1,235 @@
+
+This "protection" layer of our sensitive value is based on the composition of two AWS services, [Key Management Service](https://aws.amazon.com/kms/) is the one responsible for creating the private key which will be used to encrypt our value, then [Secret Manager](https://aws.amazon.com/es/secrets-manager/) or [System Manager: Parameters Store](https://aws.amazon.com/systems-manager/), both services allows us to accomplish the same functionality but differ on some caveats, _the chosen one_ will be used for saving the secured encrypted value, here is a quick graphic representation of the implementation:
+
+![A diagram that explains how services are used](./assets/03.image.2.png)
+
+Now lets translate these concepts into code, the first this we need after having the lambda functional and usable is to create the key used to encrypt our _precious value_...
+
+![Meme of our precious value](./assets/03.image.3.png)
+
+So lets create the key using the "root" account and give access to lambda just for `"kms:Decrypt"`, not other operations!
+
+> Please notice how we control and specify that only a specific ARN can call this value by: `kms:CallerArn: !Sub arn:aws:lambda:region:${AWS::AccountId}:function:myfunction`
+
+```yaml
+resources:
+    MyPrivateKey:
+      Type: AWS::KMS::Key
+      Properties:
+        Description: "My KMS private key"
+        KeyUsage: ENCRYPT_DECRYPT
+        KeyPolicy:
+          Version: "2012-10-17"
+          Id: "my-key-1"
+          Statement:
+            - Sid: "Enable root user permissions"
+              Effect: "Allow"
+              Principal:
+                AWS: !Sub arn:aws:iam::${AWS::AccountId}:root
+              Action: "kms:*"
+              Resource: "*"
+            - Sid: "Allow use of the key"
+              Effect: "Allow"
+              Principal:
+                Service: lambda.amazonaws.com
+              Action:
+                - "kms:Decrypt"
+              Resource: !Sub arn:aws:kms:region:${AWS::AccountId}:key/my-key-1
+              Condition:
+                StringEquals:
+                  kms:CallerArn: !Sub arn:aws:lambda:${AWS::Region}:${AWS::AccountId}:function:myfunction
+    MyPrivateKeyAlias:
+      Type: "AWS::KMS::Alias"
+      Properties:
+        AliasName: alias/myprivatekey
+        TargetKeyId: !Ref MyPrivateKey
+  Outputs:
+    # we will need this value later
+    MyKeyId:
+      Value: !Ref MyPrivateKey
+      Description: The ID of the KMS key used to encrypt
+```
+
+In this stage you should be able to have the lambda function, a secret key which only the root user and the lambda ARN can access and use it, now, how about create our encrypted values?
+
+For simplicity I personally prefer [AWS CLI](https://aws.amazon.com/cli/) to handles process of encrypting and saving instead of creating another lambda script that will need to use [AWS SDK](https://aws.amazon.com/sdk-for-javascript/), ending in the same result 🤷🏽
```

Comment:
```suggestion
For simplicity I personally prefer [AWS CLI](https://aws.amazon.com/cli/) to handle the process of encrypting and saving instead of creating another lambda script that will need to use [AWS SDK](https://aws.amazon.com/sdk-for-javascript/), ending in the same result 🤷🏽
```